packages/rest/src/coercion/coerce-parameter.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  SchemaObject,
+  ReferenceObject,
+  isReferenceObject,
+} from '@loopback/openapi-v3-types';
+
+import * as HttpErrors from 'http-errors';
+
+/**
+ * Coerce the http raw data to a JavaScript type data of a parameter
+ * according to its OpenAPI schema specification.
+ *
+ * @param data The raw data get from http request
+ * @param schema The parameter's schema defined in OpenAPI specification
+ */
+export function coerceParameter(
+  data: string,
+  schema?: SchemaObject | ReferenceObject,
+) {
+  // ignore reference schema
+  if (!schema || isReferenceObject(schema)) return data;
+  let coercedResult;
+  coercedResult = data;
+  const OAIType = getOAIPrimitiveType(schema.type, schema.format);
+
+  // Review Note: [Validation place 1] Validation rules can be applied for each case
+  switch (OAIType) {
+    case 'byte':
+      coercedResult = Buffer.from(data, 'base64');
+      break;
+    case 'date':
+      coercedResult = new Date(data);
+      break;
+    case 'float':
+    case 'double':
+      coercedResult = parseFloat(data);
+      break;
+    case 'number':
+    case 'long':
+      coercedResult = Number(data);
+      break;
+    case 'integer':
+      coercedResult = parseInt(data);
+      break;
+    case 'boolean':
+      coercedResult = isTrue(data) ? true : false;
+    case 'string':
+    case 'password':
+    // serizlize will be supported in next PR
+    case 'serialize':
+      break;
+    case 'unknownType':
+    default:
+      throw new HttpErrors.NotImplemented(
+        `Type ${schema.type} with format ${
+          schema.format
+        } is not a valid OpenAPI schema`,
+      );
+  }
+  return coercedResult;
+}
+
+/**
+ * A set of truthy values. A data in this set will be coerced to `true`,
+ * otherwise coerced to `false`.
+ *
+ * @param data The raw data get from http request
+ * @returns The corresponding coerced boolean type
+ */
+
+function isTrue(data: string): boolean {
+  const isTrueSet = ['true', '1', true, 1];
```

Comment:
are `1`, `true`, `'1'` reachable here? Or is this implemented for the purpose of using `@loopback/types` later on?

Reply:
Can we make `boolean` coercion more strict please and reject unknown values (throw an error)?

For example, both values `"TRUE"` and `"ok"` are interpreted as `false` right now, which is very confusing!

 - IMO, we should ignore case here and treat `TRUE` the same way as `true`.
 - I am proposing to reject any value that's not `true`, `false`, `1,` or `0`. 
 - An edge case to figure out: how to treat an empty string? I think we should convert it to `undefined` and let validation handle it later.

This is the behavior we have in LB 3.x, see https://github.com/strongloop/strong-remoting/blob/032e51e1f5bf69a91bc1b89ec73f734a26c29ad2/test/rest-coercion/urlencoded-boolean.suite.js#L42-L45

Reply:
@shimks @bajtos This PR would not be the full implementation for `boolean` type coercion, that's why I didn't write all test cases and handle the validations. I think we agree to create PRs incrementally add edge cases.

> I think the changes presented in this pull request are reasonable and we may take a different approach to incremental steps - start with a wide implementation that covers multiple types but does not handle all edge cases - at least that's how this pull request seems to me. I don't mind either way as long as we can iterate quickly and avoid a pull request that will take weeks to land.

And this PR won't be the end of story #750, it will include all edge cases for all types we support.

I think what I miss is @bajtos 's comment in https://github.com/strongloop/loopback-next/pull/1370#issuecomment-393081292, I should pick one type and
> start with a single type number and figure out all aspects of coercion

Will do it in the next commit.

Reply:
@shimks @bajtos FYI I will track the coming PRs in https://github.com/strongloop/loopback-next/issues/750#issuecomment-394481857

Reply:
@jannyHou sounds good! I am fine to leave out edge cases from this PR as long as we are keeping track of missing items (a todo list).

Could you please move the todo list from a comment in #750 to the issue description at the top of #750?